test/test_torch.py

```diff
@@ -1303,6 +1303,73 @@ def test_dtypes(self):
         cuda_dtypes = [d for d in all_dtypes if d.is_cuda]
         self._test_dtypes(self, cpu_dtypes, cuda_dtypes, torch.strided)
 
+    def test_device(self):
+        self.assertEqual('cpu', torch.tensor(5).device)
+        self.assertEqual('cpu', torch.ones((2, 3), dtype=torch.float32, device='cpu').device)
+        self.assertEqual('cpu', torch.ones((2, 3), dtype=torch.float32, device='cpu:0').device)
+        if torch.cuda.is_available():
+            self.assertEqual('cuda:0', torch.tensor(5).cuda(0).device)
+            self.assertEqual('cuda:0', torch.tensor(5).cuda('cuda:0').device)
+            self.assertRaises(RuntimeError, lambda: torch.tensor(5).cuda('cpu'))
+            self.assertRaises(RuntimeError, lambda: torch.tensor(5).cuda('cpu:0'))
+            self.assertEqual('cuda:0', torch.tensor(5, dtype=torch.cuda.int64, device=0).device)
+            self.assertEqual('cuda:0', torch.tensor(5, dtype=torch.cuda.int64, device='cuda:0').device)
+            self.assertEqual('cuda:' + str(torch.cuda.current_device()),
+                             torch.tensor(5, dtype=torch.cuda.int64, device='cuda').device)
+
+            if torch.cuda.device_count() > 1:
+                self.assertEqual('cuda:1', torch.tensor(5).cuda(1).device)
+                self.assertEqual('cuda:1', torch.tensor(5).cuda('cuda:1').device)
+                self.assertEqual('cuda:1', torch.tensor(5, dtype=torch.cuda.int64, device=1).device)
+                self.assertEqual('cuda:1', torch.tensor(5, dtype=torch.cuda.int64, device='cuda:1').device)
+
+    def test_devicespec(self):
+        cpu = torch.DeviceSpec('cpu')
+        self.assertEqual('cpu', str(cpu))
+        self.assertEqual('cpu', cpu.device_type)
+        self.assertEqual(None, cpu.device_index)
+        self.assertRaises(RuntimeError, lambda: cpu.cuda_device_index)
+
+        cpu0 = torch.DeviceSpec('cpu:0')
```

Comment:
I know this comment is a ton of bikeshedding, but this might save people a lot of typing:

Any reason why we call this a `DeviceSpec` and not a `Device` like the C++ struct? Also, how about making it `torch.device(...)`? That's a bit more consistent with layouts/dtypes (no uppercase stuff), and doesn't have to imply that it's actually a class constructor (e.g. can return different classes/look up singletons/etc.)

Reply:
Hmm using a string for "the device the really exists" is weird, because there are so many strings that don't represent real devices! I think it would make more sense to have our own custom type for devices that do exist, and we don't have to check that, and e.g. use strings for those unrealized cases. Wdyt?

Reply:
Well, it's the string representation of the valid Device type that doesn't exist yet :).  In all seriousness, I don't think there's really a need for a separate type yet (which is why I didn't implement it); I was more concerned with future proofing.  But it seems like we can return a DeviceSpec for now and if we need to change it in the future, make a backwards-compatible (except for explicit type checks) Device type with the same attrs/methods.

Reply:
Oh, I'm not proposing adding another type (even in the future!). I'm only suggesting that we rename `DeviceSpec` to `Device` and start treating it as a "realized device" as you called it. It would also get returned from `.device`.

Whenever you need an unrealized device, you use a string. If the string really can't be matched with any of the existing devices then it fails. That's a pretty simple scheme, and I think it is an improvement over `DeviceSpec`.

Reply:
1) Should the user be able to create a Device?
2) If 1) is true and the user passes an invalid index (for cuda), it should give an error?  So, you actually initialize cuda the first time you create a cuda Device?

Reply:
1. Not directly. The user can call `torch.device('cuda:0')`, to convert an "unrealized device" (string) to a "realized device" (`Device`, or a subclass).
2. If there's no way to get the number of devices without initializing a CUDA context, then I think so. If someone asks for a CUDA device then it seems quite reasonable to assume that they're going to use it. And if they desperately need to avoid the initialization they should just keep it as a string (which will work wherever a `Device` would).
3. Having a realized device, and not exposing the class directly has the benefit that later we could also add specialized methods for different devices. For example dropping CUDA memory cache from a single device, setting application clocks on GPUs, querying memory, etc.

Reply:
2.  My thought was this is a bit annoying; it seems reasonable to able to define a bunch of devices/devicespecs upfront and only use the ones I end up needing.  I can do this with strings, but then I can't do index math without doing the parsing myself.

Reply:
2. The same argument holds for strings returned from `.device` 😄 I don't think it's very common to do any index math on the predefined set of devices. You just put the at the top of the script and use everywhere below.

Reply:
I guess, but I like the current model that I can basically do whatever I want without cuda until I actually try to create a cuda tensor.  It's also not clear to me why users care that much whether the device is realized or not?  (we may care internally, but we can hide that fact).

Reply:
What about "lazy initializing" the device information?  That is, imagine I logically had a DeviceSpec (unrealized) and a Device (realized, with device information).  I could call these both Device and lazy initialize the information when it's accessed by realizing the Device.  We don't have device information currently, but this model would allow us to call both the user specification and the thing returned from `tensor.device` as "Device" and the former would not be initialized yet and the later would.